# Inhaltsbasiertes Empfehlungssystem für Netflix-Filme

**Gruppe 1:** <br>
Theen, Johannes (TH München)<br>
Utz, Elisabeth (OTH Amberg-Weiden)<br>
Yaruchyk, Oleg (TH München)

(Überlegungen zum Bericht)
# Textverarbeitung<br>
Um die Textdaten verarbeiten zu können, ist es notwendig, diese auf Vektoren oder Matrizen abzubilden. 

## 1. Import von Bibliotheken, Klassen und Funktionen

In [1]:
import pandas as pd
import numpy as np
import string

# Word2Vec
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.feature_extraction.text import TfidfVectorizer
# Dimensionsreduktion
from scipy import sparse
from sklearn.decomposition import TruncatedSVD

from sklearn.metrics.pairwise import cosine_similarity
from sklearn.metrics.pairwise import sigmoid_kernel
from sklearn.metrics.pairwise import polynomial_kernel

## 2. Einlesen und Analyse der Datenbank

In [2]:
raw_data = pd.read_csv('netflix_titles.csv')
print(raw_data.shape)
raw_data.head()

(6234, 12)


show_id     type                                    title  \
0  81145628    Movie  Norm of the North: King Sized Adventure   
1  80117401    Movie               Jandino: Whatever it Takes   
2  70234439  TV Show                       Transformers Prime   
3  80058654  TV Show         Transformers: Robots in Disguise   
4  80125979    Movie                             #realityhigh   

                   director  \
0  Richard Finn, Tim Maltby   
1                       NaN   
2                       NaN   
3                       NaN   
4          Fernando Lebrija   

                                                cast  \
0  Alan Marriott, Andrew Toth, Brian Dobson, Cole...   
1                                   Jandino Asporaat   
2  Peter Cullen, Sumalee Montano, Frank Welker, J...   
3  Will Friedle, Darren Criss, Constance Zimmer, ...   
4  Nesta Cooper, Kate Walsh, John Michael Higgins...   

                                    country         date_added  release_year  \
0  United States, India, South Korea, China  September 9, 2019          2019   
1                            United Kingdom  September 9, 2016          2016   
2                             United States  September 8, 2018          2013   
3                             United States  September 8, 2018          2016   
4                             United States  September 8, 2017          2017   

     rating  duration                           listed_in  \
0     TV-PG    90 min  Children & Family Movies, Comedies   
1     TV-MA    94 min                     Stand-Up Comedy   
2  TV-Y7-FV  1 Season                            Kids' TV   
3     TV-Y7  1 Season                            Kids' TV   
4     TV-14    99 min                            Comedies   

                                         description  
0  Before planning an awesome wedding for his gra...  
1  Jandino Asporaat riffs on the challenges of ra...  
2  With the help of three human allies, the Autob...  
3  When a prison ship crash unleashes hundreds of...  
4  When nerdy high schooler Dani finally attracts...

In [3]:
raw_data.describe(include='all')

show_id   type   title                director  \
count   6.234000e+03   6234    6234                    4265   
unique           NaN      2    6172                    3301   
top              NaN  Movie  Tunnel  Raúl Campos, Jan Suter   
freq             NaN   4265       3                      18   
mean    7.670368e+07    NaN     NaN                     NaN   
std     1.094296e+07    NaN     NaN                     NaN   
min     2.477470e+05    NaN     NaN                     NaN   
25%     8.003580e+07    NaN     NaN                     NaN   
50%     8.016337e+07    NaN     NaN                     NaN   
75%     8.024489e+07    NaN     NaN                     NaN   
max     8.123573e+07    NaN     NaN                     NaN   

                      cast        country       date_added  release_year  \
count                 5664           5758             6223    6234.00000   
unique                5469            554             1524           NaN   
top     David Attenborough  United States  January 1, 2020           NaN   
freq                    18           2032              122           NaN   
mean                   NaN            NaN              NaN    2013.35932   
std                    NaN            NaN              NaN       8.81162   
min                    NaN            NaN              NaN    1925.00000   
25%                    NaN            NaN              NaN    2013.00000   
50%                    NaN            NaN              NaN    2016.00000   
75%                    NaN            NaN              NaN    2018.00000   
max                    NaN            NaN              NaN    2020.00000   

       rating  duration      listed_in  \
count    6224      6234           6234   
unique     14       201            461   
top     TV-MA  1 Season  Documentaries   
freq     2027      1321            299   
mean      NaN       NaN            NaN   
std       NaN       NaN            NaN   
min       NaN       NaN            NaN   
25%       NaN       NaN            NaN   
50%       NaN       NaN            NaN   
75%       NaN       NaN            NaN   
max       NaN       NaN            NaN   

                                              description  
count                                                6234  
unique                                               6226  
top     A surly septuagenarian gets another chance at ...  
freq                                                    3  
mean                                                  NaN  
std                                                   NaN  
min                                                   NaN  
25%                                                   NaN  
50%                                                   NaN  
75%                                                   NaN  
max                                                   NaN

## 3. Preprocessing

Die Datenbank besteht aus 12 Spalten und 6.234 Zeilen. Anhand der Spalte "director" ist bereits zu erkennen, dass manche Zellen den Wert "NaN" enthalten. Die Zeile "count" in der anschließenden Ansicht macht noch einmal deutlich, dass nicht alle Spalten über 6.234 Einträge verfügen. Daher werden im nächsten Schritt alle Einträge "NaN" durch eine leere Zelle ersetzt.

In [4]:
raw_data = raw_data.fillna('')
raw_data.describe(include='all')

show_id   type   title director  cast        country  \
count   6.234000e+03   6234    6234     6234  6234           6234   
unique           NaN      2    6172     3302  5470            555   
top              NaN  Movie  Tunnel                 United States   
freq             NaN   4265       3     1969   570           2032   
mean    7.670368e+07    NaN     NaN      NaN   NaN            NaN   
std     1.094296e+07    NaN     NaN      NaN   NaN            NaN   
min     2.477470e+05    NaN     NaN      NaN   NaN            NaN   
25%     8.003580e+07    NaN     NaN      NaN   NaN            NaN   
50%     8.016337e+07    NaN     NaN      NaN   NaN            NaN   
75%     8.024489e+07    NaN     NaN      NaN   NaN            NaN   
max     8.123573e+07    NaN     NaN      NaN   NaN            NaN   

             date_added  release_year rating  duration      listed_in  \
count              6234    6234.00000   6234      6234           6234   
unique             1525           NaN     15       201            461   
top     January 1, 2020           NaN  TV-MA  1 Season  Documentaries   
freq                122           NaN   2027      1321            299   
mean                NaN    2013.35932    NaN       NaN            NaN   
std                 NaN       8.81162    NaN       NaN            NaN   
min                 NaN    1925.00000    NaN       NaN            NaN   
25%                 NaN    2013.00000    NaN       NaN            NaN   
50%                 NaN    2016.00000    NaN       NaN            NaN   
75%                 NaN    2018.00000    NaN       NaN            NaN   
max                 NaN    2020.00000    NaN       NaN            NaN   

                                              description  
count                                                6234  
unique                                               6226  
top     A surly septuagenarian gets another chance at ...  
freq                                                    3  
mean                                                  NaN  
std                                                   NaN  
min                                                   NaN  
25%                                                   NaN  
50%                                                   NaN  
75%                                                   NaN  
max                                                   NaN

Damit Schauspieler und Regisseure in die Berechnungen einbezogen werden können, werden die Vor- und Nachnamen zusammen- und alles klein geschrieben (z.B. wird Richard Finn zu richardfinn). Auch aus mehreren Wörtern bestehende Ländernamen und Kategorien werden auf diese Weise verändert.

In [5]:
# Funktion entfernt Leerzeichen, schreibt alles klein und ersetzt anschließend jedes Komma durch ein Leerzeichen 
def organize_data(data):
    data = data.str.replace(' ','')
    data = data.str.lower()
    data = data.str.replace(',',', ')
    return data

raw_data['type'] = organize_data(raw_data['type'])
raw_data['director'] = organize_data(raw_data['director'])
raw_data['cast'] = organize_data(raw_data['cast'])
raw_data['country'] = organize_data(raw_data['country'])
raw_data['listed_in'] = organize_data(raw_data['listed_in'])

raw_data.head()

show_id    type                                    title  \
0  81145628   movie  Norm of the North: King Sized Adventure   
1  80117401   movie               Jandino: Whatever it Takes   
2  70234439  tvshow                       Transformers Prime   
3  80058654  tvshow         Transformers: Robots in Disguise   
4  80125979   movie                             #realityhigh   

                 director                                               cast  \
0  richardfinn, timmaltby  alanmarriott, andrewtoth, briandobson, colehow...   
1                                                            jandinoasporaat   
2                          petercullen, sumaleemontano, frankwelker, jeff...   
3                          willfriedle, darrencriss, constancezimmer, kha...   
4         fernandolebrija  nestacooper, katewalsh, johnmichaelhiggins, ke...   

                                  country         date_added  release_year  \
0  unitedstates, india, southkorea, china  September 9, 2019          2019   
1                           unitedkingdom  September 9, 2016          2016   
2                            unitedstates  September 8, 2018          2013   
3                            unitedstates  September 8, 2018          2016   
4                            unitedstates  September 8, 2017          2017   

     rating  duration                        listed_in  \
0     TV-PG    90 min  children&familymovies, comedies   
1     TV-MA    94 min                   stand-upcomedy   
2  TV-Y7-FV  1 Season                          kids'tv   
3     TV-Y7  1 Season                          kids'tv   
4     TV-14    99 min                         comedies   

                                         description  
0  Before planning an awesome wedding for his gra...  
1  Jandino Asporaat riffs on the challenges of ra...  
2  With the help of three human allies, the Autob...  
3  When a prison ship crash unleashes hundreds of...  
4  When nerdy high schooler Dani finally attracts...

Im folgenden Schritt werden noch enthaltene (Satz-)Zeichen wie z.B. "&" in den Spalten "director", "cast", "country", "rating" und "listed_in" entfernt.

In [6]:
# Interpunktion löschen
raw_data['cast'] = [row.translate(str.maketrans("","", string.punctuation)) for row in raw_data['cast']]
raw_data['listed_in'] = [row.translate(str.maketrans("","", string.punctuation)) for row in raw_data['listed_in']]
raw_data['director'] = [row.translate(str.maketrans("","", string.punctuation)) for row in raw_data['director']]
raw_data['country'] = [row.translate(str.maketrans("","", string.punctuation)) for row in raw_data['country']]
raw_data['rating'] = [row.translate(str.maketrans("","", string.punctuation)) for row in raw_data['rating']]
raw_data.head()

show_id    type                                    title  \
0  81145628   movie  Norm of the North: King Sized Adventure   
1  80117401   movie               Jandino: Whatever it Takes   
2  70234439  tvshow                       Transformers Prime   
3  80058654  tvshow         Transformers: Robots in Disguise   
4  80125979   movie                             #realityhigh   

                director                                               cast  \
0  richardfinn timmaltby  alanmarriott andrewtoth briandobson colehoward...   
1                                                           jandinoasporaat   
2                         petercullen sumaleemontano frankwelker jeffrey...   
3                         willfriedle darrencriss constancezimmer kharyp...   
4        fernandolebrija  nestacooper katewalsh johnmichaelhiggins keith...   

                               country         date_added  release_year  \
0  unitedstates india southkorea china  September 9, 2019          2019   
1                        unitedkingdom  September 9, 2016          2016   
2                         unitedstates  September 8, 2018          2013   
3                         unitedstates  September 8, 2018          2016   
4                         unitedstates  September 8, 2017          2017   

   rating  duration                      listed_in  \
0    TVPG    90 min  childrenfamilymovies comedies   
1    TVMA    94 min                  standupcomedy   
2  TVY7FV  1 Season                         kidstv   
3    TVY7  1 Season                         kidstv   
4    TV14    99 min                       comedies   

                                         description  
0  Before planning an awesome wedding for his gra...  
1  Jandino Asporaat riffs on the challenges of ra...  
2  With the help of three human allies, the Autob...  
3  When a prison ship crash unleashes hundreds of...  
4  When nerdy high schooler Dani finally attracts...

## 4. Vektorisierung des Dokuments

Damit die Daten ausgewertet und verglichen werden können, werden sie nun mithilfe der Klasse CountVectorizer von scikit-learn in Vektoren transformiert. Auf die Spalte "description" wird zusätzlich der TF-IDF-Transformer von scikit-learn angewandt. So wird zwar nach wie vor die Häufigkeit von Wörtern innerhalb eines Dokuments betrachtet. Allerdings werden Wörter, die in jedem Dokument vorkommen (wie z.B. "is" oder "and") weniger stark gewichtet, da sie für uns keinen Mehrwert enthalten.

In [7]:
count_vectorizer= CountVectorizer()
tfidf_vectorizer = TfidfVectorizer()

final_data = pd.DataFrame(data=raw_data)

data_type = count_vectorizer.fit_transform(raw_data['type'])
data_title = count_vectorizer.fit_transform(raw_data['title'])
data_director = count_vectorizer.fit_transform(raw_data['director'])
data_cast = count_vectorizer.fit_transform(raw_data['cast'])
data_country = count_vectorizer.fit_transform(raw_data['country'])
data_date_added = count_vectorizer.fit_transform(raw_data['date_added'])
data_release_year = raw_data['release_year']
data_rating = count_vectorizer.fit_transform(raw_data['rating'])
data_duration = count_vectorizer.fit_transform(raw_data['duration'])
data_listed_in = count_vectorizer.fit_transform(raw_data['listed_in'])
data_description = tfidf_vectorizer.fit_transform(raw_data['description'])

In [8]:
#merge the data
all_matrix = sparse.hstack((data_type, data_title, data_director, data_cast, data_country, data_rating, data_listed_in, data_description), format='csr') 
all_matrix

<6234x54665 sparse matrix of type '<class 'numpy.float64'>'
	with 234240 stored elements in Compressed Sparse Row format>

In [11]:
from sklearn.preprocessing import MaxAbsScaler
all_matrix_scale = MaxAbsScaler().fit_transform(all_matrix)
all_matrix_scale

<6234x54665 sparse matrix of type '<class 'numpy.float64'>'
	with 234240 stored elements in Compressed Sparse Row format>

In [14]:
svd = TruncatedSVD(n_components = 5100, n_iter = 5)
all_matrix_svd = svd.fit(all_matrix_scale)
all_matrix_svd = svd.transform(all_matrix_svd)
explained_variance = svd.explained_variance_ratio_.sum()
print(explained_variance)

0.9488362595204387


In [15]:
# generating the cosine similarity matrix
similarity = cosine_similarity(all_matrix, all_matrix)
similarity_scale = cosine_similarity(all_matrix_scale, all_matrix_scale)
similarity_svd = cosine_similarity(all_matrix_svd,all_matrix_svd)

indices = pd.Series(final_data['title'])
indices[:5]

0    Norm of the North: King Sized Adventure
1                 Jandino: Whatever it Takes
2                         Transformers Prime
3           Transformers: Robots in Disguise
4                               #realityhigh
Name: title, dtype: object

In [17]:
def recommendations(title, cosine_sim = similarity_scale):
    
    recommended_movies = []
    
    # gettin the index of the movie that matches the title
    idx = indices[indices == title].index[0]

    # creating a Series with the similarity scores in descending order
    score_series = pd.Series(cosine_sim[idx]).sort_values(ascending = False)

    #print(score_series)
    # getting the indexes of the 10 most similar movies
    top_10_indexes = list(score_series.iloc[1:11].index)
    # populating the list with the titles of the best 10 matching movies
    for i in top_10_indexes:
        recommended_movies.append(list(raw_data['title'])[i])
        
    return recommended_movies

recommendations('Avengers: Infinity War')

['Thor: Ragnarok',
 'Black Panther',
 'Godzilla',
 'Scorpion King 5: Book of Souls',
 'Limitless',
 'Solo: A Star Wars Story',
 'Solo: A Star Wars Story (Spanish Version)',
 'War Horse',
 '9',
 'Her']